### PR TITLE
Support optional containers based on their contract

### PR DESCRIPTION
**Pull request by CameronDiver**
Change-type: minor
Signed-off-by: Cameron Diver <cameron@balena.io>

### CODE REVIEW

File: src/lib/contracts.ts [hunk 1]
```diff
@@ -75,23 +89,45 @@ export async function containerContractsFulfilled(
 	});
 
 	if (children.length === containers.length) {
-		return { valid: true, unmetServices: [] };
+		console.log(`Got no problems with this contract setup`);
```
Comment (richbayliss) [inline src/lib/contracts.ts]: debug output?

File: src/lib/contracts.ts [hunk 2]
```diff
@@ -75,23 +89,45 @@ export async function containerContractsFulfilled(
...
+					null
+				);
+			},
+		);
+
+		const nonOptional = _.filter(
+			unfulfilledServices,
+			svcName => !serviceContracts[svcName].optional,
+		);
+		const valid = nonOptional.length === 0;
```
Comment (Page-) [inline src/lib/contracts.ts]:
```suggestion
		const valid = _.some(
			unfulfilledServices,
			svcName => !serviceContracts[svcName].optional,
		);
```

File: src/lib/contracts.ts (hunk 1 above)
Comment (CameronDiver) [reply src/lib/contracts.ts]: haha yes, good catch!

File: src/lib/contracts.ts [hunk 3]
```diff
@@ -75,23 +89,45 @@ export async function containerContractsFulfilled(
...
+					return true;
 				}
-				return;
-			})
-			.filter(n => n != null)
-			.value() as string[];
-		return { valid: false, unmetServices };
+				// Did we find the contract in the generated state?
+				return (
+					_.find(children, child => _.isEqual((child as any).raw, contract)) !=
```
Comment (Page-) [inline src/lib/contracts.ts]:
```suggestion
					_.some(children, child => _.isEqual((child as any).raw, contract))
```

File: src/lib/contracts.ts [hunk 4]
```diff
@@ -53,7 +61,9 @@ export async function containerContractsFulfilled(
 
 	universe.addChildren(
 		[osContract, supervisorContract].concat(
```
Comment (Page-) [inline src/lib/contracts.ts]:
```suggestion
		[osContract, supervisorContract, ...containers.map(c => new Contract(c))]
```

File: src/lib/contracts.ts [hunk 5]
```diff
@@ -13,13 +13,21 @@ import supervisorVersion = require('./supervisor-version');
...
-	const containers = _.values(serviceContracts);
+): Promise<{
+	valid: boolean;
+	unmetServices: string[];
+	fulfilledServices: string[];
+}> {
+	const containers = _(serviceContracts)
+		.values()
+		.map('contract')
+		.filter(c => c != null)
```
Comment (Page-) [inline src/lib/contracts.ts]:
```suggestion
		.compact()
```
(and with this I think you can drop some `!`s below)

Comment (CameronDiver) [reply src/lib/contracts.ts]: nice! (the `!` dropping)

Comment (Page-) [reply src/lib/contracts.ts]: The other way to do it (eg with native filter) is with a type guard, ie
```ts
arr.filter((x): x is NonNullable<typeof x> => x != null)

File: src/lib/contracts.ts [hunk 6]
```diff
@@ -13,13 +13,21 @@ import supervisorVersion = require('./supervisor-version');
...
 	serviceContracts: ServiceContracts,
-): Promise<{ valid: boolean; unmetServices: string[] }> {
-	const containers = _.values(serviceContracts);
+): Promise<{
+	valid: boolean;
+	unmetServices: string[];
+	fulfilledServices: string[];
+}> {
+	const containers = _(serviceContracts)
+		.values()
```
Comment (Page-) [inline src/lib/contracts.ts]: There's no need for this `.values()` since lodash can map on an object directly

**Pull request merged**